firmware: BTC sign, supported locktime and sequence bytes

TheCharlatan opened the pull request:
The following pairs of locktime and sequence bytes are now supported:

locktime = 0 , nsequence = 0xffffffff
-> as before

locktime > 0, nsequence = 0xffffffff - 1
-> nsequence < 0xffffffff indicates a locktime on a transaction.

locktime > 0, nsequence = 0xffffffff - 2
-> nsequence == 0xffffffff - 2 indicates that the transaction is RBF. Since
we have the above rule on the locktime, the RBF rule only makes sense if the locktime is not zero again.

All other combinations of nsequence and locktime bytes are still not allowed. Other combinations should eventually be allowed as well to support BIP68 (nseuqnece byte definitions), BIP65
(OP_CHECKLOCKTIMEVERIFY) and BIP112 (OP_CHECKSEQUENCEVERIFY).

The relaxed constraints were chosen to accommodate for electrum's usage of RBF. The relation between locktime, nsequence and RBF is explained in this bitcoinwiki article: https://en.bitcoinwiki.org/wiki/NSequence .

I initially wrote this code for myself, but since it seems to work reasonably well, I thought I might as well post a pull request. My C is horrible, so fill your pockets with grains of salt before reviewing. At some point I am converting a `uint32` to a `const char *` with `snprintf(formatted_locktime, sizeof formatted_locktime, "\n\n\n%lu", (unsigned long)_locktime);`, which feels especially wrong.

EDIT: There is still one case that is supported with extra verification, which is when the sequence bytes are 0xffffffff and the locktime is > 0. I am not sure if this case should be rejected or not.

benma commented on src/ui/components/confirm_transaction.h:
```diff
@@ -43,4 +43,13 @@ component_t* confirm_transaction_fee_create(
+/**
+ * Creates a confirm screen.
+ */
+component_t* confirm_transaction_locktime_create(
```
move to a new file, `confirm_locktime.h`.

TheCharlatan replied:
Should I really move this? It displays only in combination with the other two functions in there and I'll have to redefine the functions internal to that file.

benma replied:
Hang on why don't you just use the already existing `workflow_confirm()`?

TheCharlatan replied:
Using it now.

benma commented on src/ui/components/confirm_transaction.h:
```diff
@@ -43,4 +43,13 @@ component_t* confirm_transaction_fee_create(
     void (*confirm_callback)(void),
     void (*cancel_callback)(void));
 
+/**
```
docstring incomplete, describe all params and the expected behavior of the component.

benma commented on src/workflow/verify_locktime.h:
```diff
@@ -0,0 +1,29 @@
+#include <stdbool.h>
+
+/**
+ * Shows a confirmation dialog to the user to confirm the total amount being
```
wrong docstring

benma commented on py/bitbox02/bitbox02/bitbox02.py:
```diff
@@ -267,9 +267,6 @@ def btc_sign(
         # pylint: disable=too-many-locals,no-member
 
-        # Reserved for future use.
-        assert version == 1 and locktime == 0
```
actually, the firmware check was not extended. i suggest to restore this here and we open a new PR for extending the version check.

TheCharlatan commented on src/apps/btc/btc_sign.c:
```diff
@@ -405,6 +412,16 @@ app_btc_sign_error_t app_btc_sign_output(
+        char formatted_locktime[100] = {0};
+        snprintf(
+            formatted_locktime, sizeof formatted_locktime, "\n\n\n%lu", (unsigned long)_locktime);
+        if (!workflow_verify_locktime(formatted_locktime, _rbf)) {
```
I'm copying from the behaviour of passing the amount to the workflow here. I did remove the newlines now.

TheCharlatan replied:
Passing `_locktime` now.

TheCharlatan commented on src/apps/btc/btc_sign.c:
```diff
@@ -306,7 +305,15 @@ app_btc_sign_error_t app_btc_sign_input(
+            _rbf = 1;
+        }
+        // sequence bytes indicate a locktime, so check that it is defined
+        else if (request->sequence == 0xffffffff - 1 && _locktime != 0) {
```
I left out the coin check, do you think we really need it? Electrum-ltc still sets the sequence bytes and locktime for example.

benma replied:
yeah we need it. you can't ask the user to confirm that they are using rbf in litecoin. 

i would even add a `_handle_rbf` bool that is true only for Bitcoin. reason:

we should think about even confirming with the user when they are *not* using rbf (at least in the expert mode i proposed in my blog post), and that should not slip into Litecoin either, so separating rbf tracking from whether it should be confirmed in the first place seems more robust. 

instead of the `_handle_rbf` var, maybe it even makes sense to `#if` guard all the rbf stuff using the `APP_BTC` macro.

TheCharlatan replied:
This seems a bit premature if we don't know yet how the expert mode will look like. How about just adding an `#if defined(APP_BTC)` guard around the `_rbf=true`?

TheCharlatan replied:
Added `#if defined(APP_BTC)` guard around the `_rbf=true`

benma commented on src/apps/btc/btc_sign.c:
```diff
@@ -433,7 +447,17 @@ app_btc_sign_error_t app_btc_sign_output(
+            // the RBF nsequence bytes only make sense in conjunction with a locktime,
+            // so verify both simultaneously
+            char formatted_locktime[100] = {0};
+            snprintf(
```
still a bit confused why you format the locktime here, but rbf inside the function. anyway, check if you can use `workflow_confirm` instead, which would cut down on a lot of code.

TheCharlatan replied:
Is there an easy way to move the confirmation in `workflow_confirm` to the top? Why is it on the bottom in the first place actually?

TheCharlatan replied:
It feels awkward to accept the locktime at the bottom and to the rest of the transaction workflow at the top.

benma replied:
i would say

```uint8_t slider_position = scrollable || longtouch ? top_slider : bottom_slider;```

can be made an argument instead in `_confirm_create()`, have `confirm_create()` and `confirm_create_scrollable()` provide the right position, and make a new one `confirm_create_top()` or sth that forces it to the top.

TheCharlatan replied:
Mmh, this is getting a bit complicated for what was before a straight forward change.

TheCharlatan replied:
I'll leave it at the bottom for now, can always improve it later.

benma commented on src/workflow/confirm_locktime.c:
```diff
@@ -0,0 +1,48 @@
+        snprintf(
+            formatted_locktime,
+            sizeof formatted_locktime,
+            "Locktime set to block:\n%lu\nTransaction is RBF",
```
This text is inaccurate, as the locktime can also specify a timestamp: https://en.bitcoin.it/wiki/NLockTime

I would still keep this, but reserve the unix timestamp formatting for the future, by adding a check here: `if (locktime > LOCKTIME_THRESHOLD) { Abort("unsupported locktime") }` with `LOCKTIME_THRESHOLD = 500000000`, and add the same check in btc_sign.c (`return _error(APP_BTC_SIGN_ERR_INVALID_INPUT)`).

TheCharlatan replied:
Added support for timestamps now.

benma commented on src/workflow/confirm_locktime.c:
```diff
@@ -0,0 +1,48 @@
+        snprintf(
+            formatted_locktime,
+            sizeof formatted_locktime,
+            "Locktime set to block:\n%lu\nTransaction is RBF",
```
Did you test this on a device?I think `%lu` might not work.

TheCharlatan replied:
Yes, also seems to work fine for a large number, like a timestamp. But as I said initially, I am very  unsure of the conversion.

benma commented on src/workflow/confirm_locktime.c:
```diff
@@ -0,0 +1,48 @@
+        snprintf(
+            formatted_locktime,
+            sizeof formatted_locktime,
+            "Locktime set to block:\n%lu\nTransaction is RBF",
```
transaction is not rbf here?

TheCharlatan replied:
Fixed.

benma commented on src/apps/btc/btc_sign.c:
```diff
@@ -433,7 +449,14 @@ app_btc_sign_error_t app_btc_sign_output(
-        // Done with outputs. Verify total and fee.
+        // Done with outputs. Verify locktime, total and fee.
+        if (_locktime > 0 && _locktime_applies) {
+            // the RBF nsequence bytes only make sense in conjunction with a locktime,
```
comment is not technically accurate, rbf makes sense independent of the locktime value.

benma commented on test/unit-test/test_btc_sign.c:
```diff
@@ -88,7 +88,6 @@ static void _test_btc_sign_init(void** state)
         .version = 1,
         .num_inputs = 1,
         .num_outputs = 1,
-        .locktime = 0,
```
why remove this?

TheCharlatan replied:
It's only used in that single test case. That said, it would be much better to actually repair whatever is breaking the test currently.

benma replied:
In that case please restore it. this is just the struct initialization for the tx that is used in all tests. Sure, `0` is the default, but I prefer explicitness.

TheCharlatan replied:
Done.

TheCharlatan commented on src/apps/btc/btc_sign.c:
```diff
@@ -306,7 +307,22 @@ app_btc_sign_error_t app_btc_sign_input(
+            return _error(APP_BTC_SIGN_ERR_INVALID_INPUT);
+        }
+        // rbf without a locktime is ill-defined
+        if (request->sequence == 0xffffffff - 2 && _locktime != 0) {
```
Consensus enforces the truth values of the following implications:
if locktime > 0 -> nsequence < MAX_INT
if nsequence < MAX_INT - 1 -> RBF
I changed the code to only check for these values. 
This makes me think whether we should let the user verify the nlocktime at all. The rbf check makes sense to an extent, but the nlocktime seems to have little actual impact on the user.

My Opinion: Maybe we should only verify this content in some future "expert" signing mode, since it is data that cannot be verified meaningfully by the hardware wallet after all. However, in light of the things that you showed me today @benma I see no reason to verify it with a normal user. 

I pushed a change to the logic now, that just asserts the above implications, and passes through otherwise.

benma replied:
> Consensus enforces the truth values of the following implications:
> if locktime > 0 -> nsequence < MAX_INT

I understood it differently. If nsequence < MAX_INT, locktime is in effect, and it can be any value, including 0.

If nsequence = MAX_INT for all inputs, locktime can be any value, but it is ignored (not compared to current block or time).

I'll have to think about not verifying the locktime with the user at all.

benma commented on src/apps/btc/btc_sign.c:
```diff
@@ -305,9 +307,17 @@ app_btc_sign_error_t app_btc_sign_input(
         return _error(APP_BTC_SIGN_ERR_STATE);
     }
-    if (request->sequence != 0xffffffff) {
+    if (request->sequence == 0xffffffff && _locktime > 0) {
```
this is wrong, locktime can be >0. if any other input has sequence < MAX INT it is compared to the blockheight or time, otherwise it can be anything but will be ignored.

TheCharlatan replied:
I'll remove the check and move the sequence check to `_locktime_applies`.

benma commented on src/apps/btc/btc_sign.c:
```diff
@@ -305,8 +307,13 @@ app_btc_sign_error_t app_btc_sign_input(
-    if (request->sequence != 0xffffffff) {
-        return _error(APP_BTC_SIGN_ERR_INVALID_INPUT);
+    if (request->sequence < 0xffffffff - 1) {
+#if defined(APP_BTC)
```
I think this has changed in https://github.com/digitalbitbox/bitbox02-firmware/pull/251

TheCharlatan commented:
Changed the PR to contain what I have discussed with @benma. The following rules are now enforced:
- nsequence bytes need to be 0xffffffff-2, 0xffffffff-1 or 0xffffffff
- nlocktime can span from 0 to to 500000000-1 , this makes sure that the BitBox is not signing a nlocktime based on median time, but rather based on block height.
- If nlocktime > 0 and the nsequence is <0xffffffff then the nlocktime is displayed to the user during transaction signing verification
- If nsequence == 0xffffffff-2 then the transaction is RBF and the user is informed on this

benma commented on src/apps/btc/btc_sign.c:
```diff
@@ -44,6 +46,8 @@ static uint32_t _num_inputs;
 static uint32_t _num_outputs;
 static uint32_t _locktime;
 static uint32_t _index;
+static bool _rbf;
```
i would also guard this with APP_BTC, so it's clear that it is not handled in LTC.

benma commented on src/workflow/confirm_locktime.c:
```diff
@@ -0,0 +1,54 @@
+#include <util.h>
+#include <workflow/confirm.h>
+
+bool workflow_confirm_locktime_rbf(uint32_t locktime, bool rbf)
```
This is still outstanding, it seems Litecoin will still show an RBF nessage.

benma commented:
In bitbox02.py, in `btc_sign()`:

> version, locktime: reserved for future use.

Needs to be updated with a more specific doc.

benma commented on src/apps/btc/confirm_locktime_rbf.c:
```diff
@@ -0,0 +1,58 @@
+    } else {
+        rbf_text = "";
+    }
+    if (locktime == 0 && rbf == rbf_disabled) {
```
it's awkward that part of the condition to show a dialog is at call site (`if (_locktime_applies || _rbf == rbf_on) {`) and part of it is here. please move both to a single place, so it's easier to follow the logic.

TheCharlatan replied:
Removed the check in confirm_locktime_rbf, added a comment toe the function explaining what is required from the input.

benma commented on src/apps/btc/confirm_locktime_rbf.c:
```diff
@@ -0,0 +1,58 @@
+            (unsigned long)locktime,
+            rbf_text);
+    } else {
+        snprintf(formatted_locktime_rbf, sizeof formatted_locktime_rbf, "\n%s", rbf_text);
```
Drop newline?

TheCharlatan replied:
I like it with the newline.

TheCharlatan replied:
Although, looks fine without the newline as well. Removing.

TheCharlatan commented on src/apps/btc/btc_sign.c:
```diff
@@ -305,9 +311,18 @@ app_btc_sign_error_t app_btc_sign_input(
+    if (request->sequence < 0xffffffff - 2) {
         return _error(APP_BTC_SIGN_ERR_INVALID_INPUT);
     }
+#if APP_BTC == 1
```
Uff, `request->unit` is stringy, so I cannot compare that. There is also bip44_coin, but it seems weird to compare against a BIP44 account code. I could add a field `bool rbf_support` to the `app_btc_coin_params_t` struct. I think it makes sense if we put feature differences between Bitcoin forks, clones and networks there.

TheCharlatan replied:
I went with putting the rbf setting in `app_btc_coin_params_t` now.

TheCharlatan commented on src/apps/btc/btc_sign.c:
```diff
@@ -16,13 +16,15 @@
 #include <workflow/verify_recipient.h>
 #include <workflow/verify_total.h>
 
+#include <stdio.h>
```
Nothing :P

TheCharlatan commented:
I squashed the commits and added two more test cases for btc_sign, but did not add a test case for checking if rbf is really disabled for litecoin yet.

TheCharlatan commented:
Pushed a separate commit for testing Litecoin with rbf and locktime > 0 now.

benma commented on test/unit-test/test_btc_sign.c:
```diff
@@ -381,6 +383,15 @@ static void _sign(const _modification_t* mod)
+        init_req.coin = BTCCoin_LTC;
+        init_req.locktime = 1;
+        inputs[0].sequence = 0xffffffff - 2;
+        inputs[0].keypath[1] += 2;
```
slight preference for `= 0 + BIP32_INITIAL_HARDENED_CHILD`, as `+=` means this value actually depends on `inputs[0].keypath[1]` (even if it changes), but it doesn't.

TheCharlatan commented on test/unit-test/test_btc_sign.c:
```diff
@@ -250,6 +254,9 @@ static void _sign(const _modification_t* mod)
     if (mod->wrong_sequence_number) {
         inputs[0].sequence = 0;
     }
+    if (mod->user_aborts_locktime_rbf) {
```
I think I am testing that already with the litecoin_rbf_disabled case.
